crates/configuration/src/network.rs

```diff
@@ -42,48 +60,114 @@ pub struct NetworkConfig {
-        todo!()
+impl NetworkConfigBuilder<NoRelayChain> {
+    pub fn new() -> Self {
+        NetworkConfigBuilder::default()
     }
 }
 
-impl NetworkConfig {
-    pub fn new() -> NetworkConfig {
-        Self::default()
+impl NetworkConfigBuilder<WithRelayChain> {
+    pub fn build(self) -> Result<NetworkConfig, ConfigError> {
+        // TODO: validate here.
+        Ok(NetworkConfig {
+            global_settings: self.global_settings,
+            relaychain:      self.relaychain,
+            parachains:      self.parachains,
+            hrmp_channels:   self.hrmp_channels,
+        })
     }
+}
 
-    pub fn with_global_settings(self, f: fn(GlobalSettings) -> GlobalSettings) -> Self {
-        Self {
+impl<R> NetworkConfigBuilder<R> {
+    pub fn with_global_settings(
+        self,
+        f: fn(GlobalSettings) -> GlobalSettings,
+    ) -> NetworkConfigBuilder<R> {
+        NetworkConfigBuilder {
             global_settings: f(self.global_settings),
             ..self
         }
     }
+}
 
-    pub fn with_relaychain(self, f: fn(RelaychainConfig) -> RelaychainConfig) -> Self {
-        Self {
-            relaychain: f(self.relaychain),
-            ..self
+impl NetworkConfigBuilder<NoRelayChain> {
+    pub fn with_relaychain(
+        self,
+        f: fn(RelaychainConfig) -> RelaychainConfig,
+    ) -> NetworkConfigBuilder<WithRelayChain> {
+        NetworkConfigBuilder {
+            relaychain:      f(RelaychainConfig::default()),
+            global_settings: self.global_settings,
+            parachains:      self.parachains,
+            hrmp_channels:   self.hrmp_channels,
+            typestate:       PhantomData,
         }
     }
+}
 
-    pub fn with_parachain(self, f: fn(ParachainConfig) -> ParachainConfig) -> Self {
+impl NetworkConfigBuilder<WithRelayChain> {
+    pub fn with_parachain(
+        self,
+        f: fn(ParachainConfig) -> ParachainConfig,
+    ) -> NetworkConfigBuilder<WithRelayChain> {
         Self {
             parachains: vec![self.parachains, vec![f(ParachainConfig::default())]].concat(),
             ..self
         }
     }
 
-    pub fn with_hrmp_channel(self, f: fn(HrmpChannelConfig) -> HrmpChannelConfig) -> Self {
+    pub fn with_hrmp_channel(
+        self,
+        f: fn(HrmpChannelConfig) -> HrmpChannelConfig,
+    ) -> NetworkConfigBuilder<WithRelayChain> {
         Self {
             hrmp_channels: vec![self.hrmp_channels, vec![f(HrmpChannelConfig::default())]].concat(),
             ..self
         }
     }
 }
+
+impl NetworkConfig {
+    pub(crate) fn global_settings(&self) -> &GlobalSettings {
+        &self.global_settings
+    }
+
+    pub(crate) fn relaychain(&self) -> &RelaychainConfig {
+        &self.relaychain
+    }
+
+    pub(crate) fn parachains(&self) -> &Vec<ParachainConfig> {
+        &self.parachains
+    }
+
+    pub(crate) fn hrmp_channels(&self) -> &Vec<HrmpChannelConfig> {
+        &self.hrmp_channels
+    }
+
+    //[TODO]: skill serializing empty vec or None
```

Comment:
I don't understand the comment here

Reply:
We need to mark the fields to `skip if...` in serde to not not include `null`s in the json output (some empty vecs can also be skipped)

```json
{
  "global_settings": {
    "bootnode_address": [],
    "network_spawn_timeout": 1000,
    "node_spawn_timeout": 300,
    "local_ip": null
  },
  "relaychain": {
    "default_command": "polkadot",
    "default_image": "parity/polkadot:latest",
    "default_resources": null,
    "default_db_snapshot": null,
    "chain": "rococo-local",
    "chain_spec_path": null,
    "default_args": [],
    "random_nominators_count": null,
    "max_nominations": null,
    "nodes": []
  },
  "parachains": [
    {
      "id": 100,
      "chain": null,
      "registration_strategy": null,
      "initial_balance": 1000000000,
      "genesis_wasm_path": null,
      "genesis_wasm_generator": null,
      "genesis_state_path": null,
      "genesis_state_generator": null,
      "chain_spec_path": null,
      "is_cumulus_based": true,
      "bootnodes_addresses": [],
      "collators": []
    }
  ],
  "hrmp_channels": []
}
```